File: Javascript/DP/H-1289. Minimum Falling Path Sum II.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minFallingPathSum = function (grid) {
  const n = grid.length;
  const minPaths = new Array(n).fill(null).map((_) => new Array(n).fill(0));
  let lastRow1stMinNumIdx = -1;
  let lastRow2ndMinNumIdx = -1;
  /** init minPaths[0], lastRow1stMinNumIdx, and lastRow2ndMinNumIdx */
  for (let i = 0; i < n; i++) {
    const val = grid[0][i];
    minPaths[0][i] = val;
    if (
      val <
      (lastRow1stMinNumIdx === -1 ? Infinity : grid[0][lastRow1stMinNumIdx])
    ) {
      lastRow2ndMinNumIdx = lastRow1stMinNumIdx;
      lastRow1stMinNumIdx = i;
    } else if (
      val <
      (lastRow2ndMinNumIdx === -1 ? Infinity : grid[0][lastRow2ndMinNumIdx])
    ) {
      lastRow2ndMinNumIdx = i;
    }
  }
  /** start traveling grid */
  for (let i = 1; i < n; i++) {
    let curRow1stMinNumIdx = -1;
    let curRow2ndMinNumIdx = -1;
    for (let j = 0; j < n; j++) {
      const val = grid[i][j];
      if (j !== lastRow1stMinNumIdx) {
        minPaths[i][j] = val + minPaths[i - 1][lastRow1stMinNumIdx];
      } else {
        minPaths[i][j] = val + minPaths[i - 1][lastRow2ndMinNumIdx];
      }
      /** find current row's 2 smallest number's index */
      if (
        minPaths[i][j] <
        (curRow1stMinNumIdx === -1 ? Infinity : minPaths[i][curRow1stMinNumIdx])
      ) {
        curRow2ndMinNumIdx = curRow1stMinNumIdx;
        curRow1stMinNumIdx = j;
      } else if (
        minPaths[i][j] <
        (curRow2ndMinNumIdx === -1 ? Infinity : minPaths[i][curRow2ndMinNumIdx])
      ) {
        curRow2ndMinNumIdx = j;
      }
    }
    lastRow1stMinNumIdx = curRow1stMinNumIdx;
    lastRow2ndMinNumIdx = curRow2ndMinNumIdx;
  }

  return Math.min(...minPaths[n - 1]);
};

const testData = [
  {
    data: [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ],
    ans: 13, //1, 5, 7
  },
  {
    data: [[7]],
    ans: 7,
  },
  {
    data: [
      [2, 2, 1, 2, 2],
      [2, 2, 1, 2, 2],
      [2, 2, 1, 2, 2],
      [2, 2, 1, 2, 2],
      [2, 2, 1, 2, 2],
    ],
    ans: 7,
  },
];

for (const data of testData) {
  console.log(data.ans === minFallingPathSum(data.data));
}

File: DP/H-1289. Minimum Falling Path Sum II/H-1289. Minimum Falling Path Sum II.py
import sys


class Solution:
    @classmethod
    def minFallingPathSum(self, grid) -> int:
        n = len(grid)
        min_paths = [[0 for j in range(n)] for i in range(n)]
        max_safe_int = sys.maxsize
        last_row_1st_min_num_idx = -1
        last_row_2nd_min_num_idx = -1
        # init min_paths[0], last_row_1st_min_num_idx, and last_row_2nd_min_num_idx
        for i in range(n):
            val = grid[0][i]
            min_paths[0][i] = val
            if (val < (max_safe_int if last_row_1st_min_num_idx == -1 else grid[0][last_row_1st_min_num_idx])):
                last_row_2nd_min_num_idx = last_row_1st_min_num_idx
                last_row_1st_min_num_idx = i
            elif (val < (max_safe_int if last_row_2nd_min_num_idx == -1 else grid[0][last_row_2nd_min_num_idx])):
                last_row_2nd_min_num_idx = i

        # start travaling
        for i in range(1, n):
            cur_row_1st_min_num_idx = -1
            cur_row_2nd_min_num_idx = -1
            for j in range(n):
                val = grid[i][j]
                if j != last_row_1st_min_num_idx:
                    min_paths[i][j] = val + \
                        min_paths[i - 1][last_row_1st_min_num_idx]
                else:
                    min_paths[i][j] = val + \
                        min_paths[i - 1][last_row_2nd_min_num_idx]
                # update current row's 2 smallest number index
                if (min_paths[i][j] < (max_safe_int if cur_row_1st_min_num_idx == -1 else min_paths[i][cur_row_1st_min_num_idx])):
                    cur_row_2nd_min_num_idx = cur_row_1st_min_num_idx
                    cur_row_1st_min_num_idx = j
                elif (min_paths[i][j] < (max_safe_int if cur_row_2nd_min_num_idx == -1 else min_paths[i][cur_row_2nd_min_num_idx])):
                    cur_row_2nd_min_num_idx = j
            last_row_1st_min_num_idx = cur_row_1st_min_num_idx
            last_row_2nd_min_num_idx = cur_row_2nd_min_num_idx
        return min(min_paths[n-1])


test_data = [
    {
        "data": [
            [1, 2, 3],
            [4, 5, 6],
            [7, 8, 9],
        ],
        "ans": 13,  # 1, 5, 7
    },
    {
        "data": [[7]],
        "ans": 7,
    },
    {
        "data": [
            [2, 2, 1, 2, 2],
            [2, 2, 1, 2, 2],
            [2, 2, 1, 2, 2],
            [2, 2, 1, 2, 2],
            [2, 2, 1, 2, 2],
        ],
        "ans": 7,
    },
    {
        "data": [
            [-37, 51, -36, 34, -22],
            [82, 4, 30, 14, 38],
            [-68, -52, -92, 65, -85],
            [-49, -3, -77, 8, -19],
            [-60, -71, -21, -62, -73]
        ],
        "ans": -268,
    }
]

for data in test_data:
    print(data["ans"] == Solution.minFallingPathSum(data["data"]))
